### PR TITLE
Handle the case when `watch.exitCode` may return null

### PR DESCRIPTION
**Pull request by kuzaxak**
After changes introduced in https://github.com/jenkinsci/kubernetes-plugin/pull/1260 we have frequent error with `NullPointerException`. Based on a [documentation][2] for `watch.exitCode` method it's clear that watch.exitCode() can return null if close is received before the exit code. In such a case, you can handle the null value and provide an appropriate default or error value.

```
java.lang.NullPointerException at
org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecProc.join(ContainerExecProc.java:89)
at hudson.Proc.joinWithTimeout(Proc.java:174)
at com.cloudbees.jenkins.plugins.sshagent.exec.ExecRemoteAgent.stop(ExecRemoteAgent.java:129)
```

Fix for JENKINS-71135

[2]: https://github.com/fabric8io/kubernetes-client/blob/8a879128d893627a40d64cf2a9c7d5da2e7f47d2/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/dsl/ExecWatch.java#L68-L82

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 1]
```diff
@@ -86,7 +87,14 @@ public int join() throws IOException, InterruptedException {
             LOGGER.log(Level.FINEST, "Waiting for websocket to close on command finish ({0})", finished);
             finished.await();
             LOGGER.log(Level.FINEST, "Command is finished ({0})", finished);
-            return watch.exitCode().join();
+
+            CompletableFuture<Integer> exitCodeFuture = watch.exitCode();
+            if (exitCodeFuture == null) {
+                LOGGER.log(Level.FINEST, "Watcher return 'null' instead of exitCode");
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:93]: Maybe we should extend https://github.com/jenkinsci/kubernetes-plugin/blob/294ad4250b9169313db589c44fe993f5191e7ce8/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java#L563-L567 to pass the `printStream` to `proc` so unusual issues can be printed in the build log, rather than the system log where they would only be visible to admins.

**Review by jglick — APPROVED** [review]

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java (hunk 1 above)
Comment (kuzaxak) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:93]: Good idea, but I don't know how to do it in a better way. Can I add an additional arg for `ContainerExecProc` to pass `printStream` and then use it to print a log?

Comment (kuzaxak) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:93]: I think it can be implemented as a separate PR to unblock the issue fix. I can prepare a PR for that.

Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:93]: > Can I add an additional arg

That is what I was thinking.

> I think it can be implemented as a separate PR

Agreed, it was just an idea.

**Comment by annaagaf** [comment]
We have been running this build and it's still throwing NPEs at `exitCodeFuture.join() `:

```
10:31:47  java.lang.NullPointerException
10:31:47  	at org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecProc.join(ContainerExecProc.java:97)
10:31:47  	at hudson.Proc.joinWithTimeout(Proc.java:174)
10:31:47  	at com.cloudbees.jenkins.plugins.sshagent.exec.ExecRemoteAgent.stop(ExecRemoteAgent.java:129)
```

**Comment by annaagaf** [comment]
@jglick Added a commit to address another NullPointerException issue. We will also be testing this build on our Jenkins.

**Comment by annaagaf** [comment]
The updated version is still throwing NPEs, back to the drawing board.
```
10:21:32  java.lang.NullPointerException
10:21:32  	at org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecProc.join(ContainerExecProc.java:99)
10:21:32  	at hudson.Proc.joinWithTimeout(Proc.java:174)
```

**Review by jglick — DISMISSED** [review]

**Comment by annaagaf** [comment]
Updated this PR to the current state we are running. This solves NPEs and improves logging; however, this now shows the original issue, which is with ssh-agent plugin (`ssh-agent -k `failing at the end of pipeline). Will be patching that plugin to validate.

File: pom.xml [hunk 2]
```diff
@@ -3,7 +3,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>4.61</version>
+    <version>4.62</version>
```
Comment (Vlatombe) [inline pom.xml:6]: Please, for next time, avoid introducing unrelated changes, as it makes the review work harder and increases the likelihood of merge conflicts.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 3]
```diff
@@ -34,29 +37,32 @@ public class ContainerExecProc extends Proc implements Closeable, Runnable {
...
+        this(watch, alive, finished, stdin, (PrintStream) null);
     }
 
-    public ContainerExecProc(ExecWatch watch, AtomicBoolean alive, CountDownLatch finished, OutputStream stdin) {
+    public ContainerExecProc(ExecWatch watch, AtomicBoolean alive, CountDownLatch finished, OutputStream stdin, PrintStream printStream) {
         this.watch = watch;
         this.stdin = stdin == null ? watch.getInput() : stdin;
         this.alive = alive;
         this.finished = finished;
+        this.printStream = printStream;
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:65]: To avoid any NPE or null check in usage of this field.

```suggestion
        this.printStream = printStream == null ? NullPrintStream.NULL_PRINT_STREAM : printStream;
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 4]
```diff
@@ -86,7 +92,24 @@ public int join() throws IOException, InterruptedException {
             LOGGER.log(Level.FINEST, "Waiting for websocket to close on command finish ({0})", finished);
             finished.await();
             LOGGER.log(Level.FINEST, "Command is finished ({0})", finished);
-            return watch.exitCode().join();
+
+            CompletableFuture<Integer> exitCodeFuture = watch.exitCode();
+            Integer exitCode = exitCodeFuture.get();
+
+            if (exitCode == null) {
+                LOGGER.log(Level.FINEST, "Watcher return 'null' instead of exitCode");
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:100]: I don't think it is useful to mention `null` in a user facing log message if we can better diagnose it,
```suggestion
                LOGGER.log(Level.FINEST, "The container exec watch was closed before it could obtain an exit code from the process.");
```
The same message should be used as well for what gets sent to the print stream.

**Comment by mbecca** [comment]
Hello, does anyone have a date to include this bug fix?

File: pom.xml (hunk 2 above)
Comment (anna-agafonova) [reply pom.xml]: Thanks for the note! Currently rebased to latest upstream and this bump is no longer included with PR.

**Comment by anna-agafonova** [comment]
@Vlatombe @jglick addressed the PR comments, please re-review

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 5]
```diff
@@ -1,25 +1,28 @@
 package org.csanchez.jenkins.plugins.kubernetes.pipeline;
 
 
+import hudson.Proc;
+import io.fabric8.kubernetes.client.dsl.ExecWatch;
+import io.jenkins.cli.shaded.org.apache.commons.io.output.NullPrintStream;
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:6]:
```suggestion
import org.apache.commons.io.output.NullPrintStream;
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 6]
```diff
@@ -86,7 +92,24 @@ public int join() throws IOException, InterruptedException {
             LOGGER.log(Level.FINEST, "Waiting for websocket to close on command finish ({0})", finished);
             finished.await();
             LOGGER.log(Level.FINEST, "Command is finished ({0})", finished);
-            return watch.exitCode().join();
+
+            CompletableFuture<Integer> exitCodeFuture = watch.exitCode();
+            Integer exitCode = exitCodeFuture.get();
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:97]: Does this not contradict the PR title, that it is possible `exitCodeFuture == null`? Either the code or the PR title is incorrect.

Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:97]: (Docs are ambiguous whether `null` refers to `CompletableFuture<Integer>`, or the value of the `CompletableFuture`.)

Comment (anna-agafonova) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:97]: The goal here is to handle all possible outcomes (null `exitCode` is handled in this block, null `exitCodeFuture` in catchall block, checked `ExecutionException` that can be thrown by `get() ` in catch block). 
The PR title refers to the original issue that's causing NPEs on current master when `exitCode` is null, but what would you suggest as a better naming/solution here?

Comment (jglick) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:97]: > null `exitCodeFuture` in catchall block

If you mean in the `catch (Exception` block, then yes, but code should never knowingly be catching `NullPointerException`; it should check for null. Conversely, if you were relying on arbitrary unknown exceptions, you might as well keep it simple

```java
try {
    return watch.exitCode().join();
} catch (Exception x) {
    LOGGER.log(Level.FINEST, "Exception occurred while waiting for exit code", x);
    return -1;
}
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java [hunk 7]
```diff
@@ -86,7 +92,24 @@ public int join() throws IOException, InterruptedException {
...
+                return -1;
+            }
+            return exitCode;
+        } catch (ExecutionException e) {
+            LOGGER.log(Level.FINEST, "ExecutionException occurred while waiting for exit code", e.getCause());
+            printStream.printf("ExecutionException occurred while waiting for exit code: %s%n", e.getCause());
+            return -1;
+        } catch (Exception e) {
+            LOGGER.log(Level.FINEST, "Exception occurred while waiting for exit code", e.getCause());
+            printStream.printf("Exception occurred while waiting for exit code: %s: %s%n", e.getClass(), e.getCause());
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:111]:
```suggestion
            LOGGER.log(Level.FINEST, "Exception occurred while waiting for exit code", e);
            printStream.printf("Exception occurred while waiting for exit code: %s%n", e);
```
or use `Functions.printStackTrace`

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java (hunk 6 above)
Comment (anna-agafonova) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:104]: Alright, agreed - how about now?

Comment (anna-agafonova) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecProc.java:104]: @jglick could you please re-review?

**Review by Vlatombe — APPROVED** [review]
